kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,782 @@
+      preferred_ip_type = qtype
+    end
+  end
+
+  preferred_ip_type = preferred_ip_type or TYPE_A
+
+  -- parse hosts
+  local hosts = init_hosts(cache, opts.hosts, preferred_ip_type)
+
+  return setmetatable({
+    cache         = cache,
+    stats         = {},
+    hosts         = hosts,
+    r_opts        = r_opts,
+    resolv        = opts._resolv or resolv,
+    valid_ttl     = opts.valid_ttl,
+    error_ttl     = opts.error_ttl or DEFAULT_ERROR_TTL,
+    stale_ttl     = opts.stale_ttl or DEFAULT_STALE_TTL,
+    empty_ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    search_types  = search_types,
+    -- quickly accessible constant empty answers
+    empty_answers = {
+      errcode = EMPTY_RECORD_ERROR_CODE,
+      errstr  = EMPTY_RECORD_ERROR_MESSAGE,
+      ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    },
+  }, mt)
+end
+
+
+local function process_answers(self, qname, qtype, answers)
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = errcode == NAME_ERROR_CODE and self.empty_ttl or self.error_ttl
+    return answers
+  end
+
+  local processed_answers = {}
+  local cname_answer
+
+  local ttl = math_min(self.valid_ttl or 0xffffffff, 0xffffffff)  -- 0xffffffff for maximum TTL value
+
+  for _, answer in ipairs(answers) do
+    answer.name = string_lower(answer.name)
+
+    if answer.type == TYPE_CNAME then
+      cname_answer = answer   -- use the last one as the real cname
+
+    elseif answer.type == qtype then
+      -- compatible with balancer, see https://github.com/Kong/kong/pull/3088
+      if answer.type == TYPE_AAAA then
+        answer.address = ipv6_bracket(answer.address)
+
+      elseif answer.type == TYPE_SRV then
+        answer.target = ipv6_bracket(answer.target)
+      end
+
+      -- skip the SRV record pointing to itself,
+      -- see https://github.com/Kong/lua-resty-dns-client/pull/3
+      if not (answer.type == TYPE_SRV and answer.target == qname) then
+        table_insert(processed_answers, answer)
+      end
+    end
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+  end
+
+  if #processed_answers == 0 then
+    if not cname_answer then
+      return self.empty_answers
+    end
+
+    processed_answers[1] = cname_answer
+
+    log(DEBUG, "processed cname:", cname_answer.cname)
+
+  else
+    log(DEBUG, "processed ans:", #processed_answers)
+  end
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
+
+  return processed_answers
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+  local key = name .. ":" .. qtype
+  stats_count(self.stats, key, "query")
+
+  local r, err = resolver:new(self.r_opts)
+  if not r then
+    return nil, "failed to instantiate the resolver: " .. err
+  end
+
+  local start_time = now()
+
+  local options = { additional_section = true, qtype = qtype }
+  local answers, err = r:query(name, options)
+  if r.destroy then
```

Comment:
I think that the new `lua-resty-dns-resolver` always has method `destroy` since 0.23, do we need to backport this PR to old kong version?

Reply:
done, removed the checking for r.destroy